src/web/auth/jwt.rs

```diff
@@ -0,0 +1,368 @@
+use std::env;
+
+use jsonwebtoken::{Algorithm, Validation};
+use serde::{Deserialize, Serialize};
+
+use super::{Error, Permission, Result, Role, Validator};
+
+const SUPPORTED_ALGORITHMS: [Algorithm; 3] = [Algorithm::RS256, Algorithm::RS384, Algorithm::RS512];
+
+/// JSON Web Key. A cryptographic key used to validate JSON Web Tokens.
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+struct Key {
+    alg: Algorithm,
+    kty: String,
+    r#use: String,
+    n: String,
+    e: String,
+    kid: String,
+    x5t: String,
+    x5c: Vec<String>,
+}
+
+/// JSON Web Key Set. A collection of JSON Web Keys used by a given service.
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+struct Jwks {
+    keys: Vec<Key>,
+}
+
+impl Jwks {
+    /// Looks up the key by id, or returns None if it's not present in the key
+    /// set.
+    pub fn find(&self, key_id: &str) -> Option<&Key> {
+        self.keys.iter().find(|key| key.kid == key_id)
+    }
+}
+
+/// A set of values encoded in a JWT that the issuer claims to be true.
+#[derive(Debug, Deserialize)]
+struct Claims {
+    /// Audience (who or that the token is intended for). E.g. "https://api.xsnippet.org".
+    aud: String,
+    /// Issuer (who created and signed this token). E.g. "https://xsnippet.eu.auth0.com/".
+    iss: String,
+    /// Subject (whom the token refers to). E.g. "spam@eggs.foo".
+    sub: String,
+    /// Expiration time (seconds since Unix epoch).
+    exp: usize,
+    /// Subject permissions (e.g. vec!["admin"])
+    permissions: Vec<Permission>,
+}
+
+impl Jwks {
+    /// Returns a Jwks retrieved from the location identified by the given URI.
+    pub fn from_uri(uri: &str) -> Result<Self> {
+        let load_err = Error::Configuration(format!("Can't load Jwks state from {}", uri));
+        let json = match uri.split_once("://") {
+            Some(("http", _)) | Some(("https", _)) => reqwest::blocking::get(uri)
```

Comment:
I wonder, should we trust `http://`? I mean, in case of MITM attack, the whole system is compromised. Is it not?

Reply:
You are absolutely right. The reason it is implemented this way at the moment is so that we can serve our own JWKS via HTTP in tests and not have to deal with the intricacies of the HTTPS setup. *However*, I've also added support for loading the JWKS state for the local filesystem, so we could use that instead and forbid HTTP completely.

Another alternative would be using conditional compilation to allow HTTP URLs in debug builds and forbid them in release builds.

Reply:
I've gone for the former for the time being, as it allows us to simplify the test fixture a bit. Also, it means that there is no difference between a debug and a release build of the application (in terms of behavior).